optaplanner-examples/src/main/java/org/optaplanner/examples/meetingscheduling/optional/score/MeetingSchedulingConstraintProvider.java

```diff
@@ -0,0 +1,224 @@
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+                roomConflict(constraintFactory),
+                avoidOvertime(constraintFactory),
+                requiredAttendanceConflict(constraintFactory),
+                requiredRoomCapacity(constraintFactory),
+                startAndEndOnSameDay(constraintFactory),
+                requiredAndPreferredAttendanceConflict(constraintFactory),
+                preferredAttendanceConflict(constraintFactory),
+                doMeetingsAsSoonAsPossible(constraintFactory),
+                oneBreakBetweenConsecutiveMeetings(constraintFactory),
+                overlappingMeetings(constraintFactory),
+                assignLargerRoomsFirst(constraintFactory),
+                roomStability(constraintFactory)
+        };
+    }
+
+    // ************************************************************************
+    // Hard constraints
+    // ************************************************************************
+
+    protected Constraint roomConflict(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUnfiltered(MeetingAssignment.class)
+                .filter(leftAssignment -> leftAssignment.getRoom() != null)
+                .join(MeetingAssignment.class,
+                        equal(MeetingAssignment::getRoom, MeetingAssignment::getRoom),
+                        lessThan(MeetingAssignment::getId, MeetingAssignment::getId),
+                        filtering((leftAssignment, rightAssignment) -> rightAssignment.calculateOverlap(leftAssignment) > 0))
+                .penalizeConfigurable("Room conflict",
+                        (leftAssignment, rightAssignment) -> rightAssignment.calculateOverlap(leftAssignment));
+    }
+
+    protected Constraint avoidOvertime(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUnfiltered(MeetingAssignment.class)
+                .filter(meetingAssignment -> meetingAssignment.getStartingTimeGrain() != null)
+                .ifNotExists(TimeGrain.class,
+                        equal(MeetingAssignment::getLastTimeGrainIndex, TimeGrain::getGrainIndex))
+                .penalizeConfigurable("Don't go in overtime", MeetingAssignment::getLastTimeGrainIndex);
+    }
+
+    protected Constraint requiredAttendanceConflict(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(RequiredAttendance.class)
+                .join(RequiredAttendance.class,
+                        equal(RequiredAttendance::getPerson, RequiredAttendance::getPerson),
+                        lessThan(RequiredAttendance::getId, RequiredAttendance::getId))
+                .join(MeetingAssignment.class,
+                        filtering((leftRequiredAttendance, rightRequiredAttendance,
+                                leftAssignment) -> leftRequiredAttendance.getMeeting() == leftAssignment.getMeeting()),
```

Comment:
How does this work on `TriConstraintStreams`? I'm guessing something like: `equal(RequiredAttendance::getMeeting, MeetingAssignment::getMeeting)`. But in this case wouldn't the `RequiredAttendance` being compared be ambiguous, since there are two of them joined in this stream?

Reply:
The joiner will always take all the arguments from the left, and the single argument from the right.
So this would be something like `equal((lra, rra) -> lra.getMeeting(), MeetingAssignment::getMeeting)`.